### DISCUSSION TITLE
File transfer over SSH? ZMODEM support?

### DISCUSSION THREAD

**Discussion by dev-zero** [wezterm/wezterm#1450] [Ideas]
Not sure a terminal emulator is the right place for this, but since there's an SSH client integrated it might be:

I am occasionally logged in via SSH to some machine (possibly via a number of jumps) and would like to transfer files (mostly fetching a file from the machine I am logged in). At the moment this either needs a second tab with an SCP/SFTP session, or curl-pasting the file to some service. I am currently experimenting with forwarding a socket to a locally running webserver which receives the files, but since wezterm is anyway interpreting the character stream one could also think about implementing something like the [ZMODEM sx, sb, sz](https://manpages.ubuntu.com/manpages/bionic/man1/sz.1.html) or is that already available and I haven't found it?
It seems that [Tabby](https://tabby.sh/) implements it.

**Comment by wez** [comment]
I'm not sure about making a full blown SCP client in wezterm, as it feels like a very large amount of UI that isn't 100% aligned to being a terminal emulator, but your comment about zmodem: wezterm supports the [iTerm2 image protocol](https://iterm2.com/documentation-images.html) which is really a file transfer mechanism under the covers, although I wouldn't recommend it for really large files.

If we can agree to a limited UX around the receiving side, something like:

* Always save to a configurable downloads folder (defaulting to the typical Downloads folder for the user)
* Don't prompt for a filename.
* Use the filename from the server side if it wouldn't conflict
* Else, stick a unique-ifying numeric suffix on the end to avoid conflict
* Use a toast notification to announce status so that the remote host isn't silently filling up the local downloads folder

then teaching wezterm to do that for non-inline iterm images feels doable (you'd need https://iterm2.com/utilities/it2dl on the remote host), and might lay some foundation for potentially supporting zmodem later.

zmodem is a lot more complex though: I think it might be as invasive as the [tmux-cc](https://github.com/wez/wezterm/pull/1090) PR in terms of making it work transparently in the terminal.

**Reply by wez** [reply]
I hooked up iTerm2 file transfers in 5c222cfcdc758286b3b2766a613e610ada4beee8
Please give it a try!

**Reply by dev-zero** [reply]
It works perfectly. This is awesome, wez! Thanks a lot!

**Reply by 3288103265** [reply]
@wez  it2dl works fun, but i can't use it2ul to upload file. Is it2ul available?

**Reply by 3288103265** [reply]
I mean wezterm on windows

**Reply by wez** [reply]
the iterm2 image protocol can only transfer in one direction: downloading to the client

**Reply by fengjiawang** [reply]
It works perfectly when using "wezterm ssh". However, it fails when using "wezterm connect". Is there a way to fix it? Thanks

**Reply by wez** [reply]
@fengjiawang please file an issue for that

**Reply by JasonChen89** [reply]
Itrem2 natively does not support zmodem too.
But its triggers can be extended to use zmodem。
So I think what wezterm needs to do is just adding a feature which can trigger silent coprocss.

https://iterm2.com/triggers.html
https://github.com/kuoruan/iterm2-zmodem

**Reply by YiuTerran** [reply]
currently, I use `trzsz` instead.

**Reply by expnn** [reply]
I have written a plugin ([wezrs](https://github.com/expnn/wezrs)) that allows the use of trzsz for file transfers within wezterm. This plugin supports usage within an already opened mux domain, eliminating the need to log in using `trzsz ssh`.  I believe this solves @fengjiawang's problem. 

Alright, I was lying. Actually, I achieve this by utilizing shell to set a `User Var`, then listening for the `user-var-changed` event in wezterm, and automatically creating a tab. In this tab, I initiate `trzsz ssh` login and submit the file transfer command in the login window.

Currently, this plugin might not be robust enough, as any step in the aforementioned process could go wrong. However, I haven't found a way to check for execution failures in the existing Lua API. Therefore, any errors are silently ignored, which may lead to the execution of unexpected commands. For example:

- I need to use `window:perform_action()` to execute KeyAssignment, but I can't check if the execution was successful. This is reasonable in scenarios where shortcuts are used because we don't want to give feedback to the user every time a shortcut is pressed. However, when writing automation scripts, we want to get a return status to determine the next step.
- I use `pane:send_text()` to send the command to the pane. Since it always returns nil, I cannot check if the command was executed correctly. Additionally, this command seems to be asynchronous, so I can't check when the command has actually finished executing. As a result, the current plugin cannot automatically close the newly created tab. I suspect that the original intention of this function was not to execute commands in the pane, but merely to send a piece of text to it. It just so happens that when the sent text is a command, that command can be executed in the pane.

Below are videos I recorded of the usage:

- [Upload](https://youtu.be/8jNuF_gvy0w):
   ![wezrs-upload](https://github.com/wez/wezterm/assets/18525479/0949a554-2517-4b8c-b490-41201ab75761)

- [Download](https://youtu.be/N5jy9dKO2-0):
   ![wezrs-download](https://github.com/wez/wezterm/assets/18525479/51d677ae-3f30-4b42-8ed7-9935c4ead76e)


P.S. I hope that wezterm can provide more powerful interfaces to make implementing plugins like this easier. @wez
